modules/ingest-attachment/src/main/java/org/elasticsearch/ingest/attachment/TikaImpl.java

```diff
@@ -149,6 +149,8 @@ static PermissionCollection getRestrictedPermissions() {
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
+        // log4j needs to be able to check the environment
+        perms.add(new RuntimePermission("getenv.*"));
```

Comment:
Are there specific environment variables that log4j needs which we can give access to?

Reply:
I see `log4j.unbox.ringbuffer.size`, `log4j2.StatusLogger.level`, `log4j2.status.entries`, `log4j2.StatusLogger.DateFormat`, and `log4j2.StatusLogger.properties`, but I'm guessing there are more.

Reply:
From what I can tell though, we can't do wildcards like `"getenv.log4j*"`. Is there a lot of risk in just allowing `"getenv.*"`? I'm worried we'd miss something if we listed them all out.

Reply:
Unfortunately, within the bowels of log4j2, they're calling `System.getenv().containsKey(key);`. That `System.getenv()` needs `getenv.*`. So even if we did have the list of all the keys they're going to access, it still wouldn't be workaround as an allowlist because of this other call.

Or at least that's how it looks to me.

Reply:
Do we actually need this doPrivileged block here? I'm not really familiar with why it exists. Can we just move the permissions (not including `getenv.*`) into the module's plugin-security.policy file, and call this like an ordinary method? I tried that out and it seems to work. I'm not sure if this is just old code that predates the plugin-security.policy work, or if there is something really different it is trying to do.

Reply:
I guess the advantage of this doPrivileged block is that it only applies to Tika.parseToString(). But there's not a whole lot else going on in this module anyway.

Reply:
The point of all the permissions here are to limit what Tika has access to. Parsing code is notorious for (unintentionally) allowing shady things to occur. Instead of running with permissions of the server (which all plugins get), we drop down to no permissions, and add the explicit ones Tika needs. I think we should try to avoid allowing access to all env vars. From the stack traces Joe showed me, it looks like we need some slf4j initialization such that calls can be interjected with doPrivileged.

Reply:
It's not just slf4j though -- some parts of tika call log4j directly. Or maybe I misunderstood what you said?

Reply:
Something's not working as expected then. We're seeing this:
```
java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "getenv.*")
    at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:485)
    at java.base/java.security.AccessController.checkPermission(AccessController.java:1068)
    at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:411)
    at java.base/java.lang.System.getenv(System.java:1198)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.util.EnvironmentPropertySource.containsProperty(EnvironmentPropertySource.java:99)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.util.PropertiesUtil$Environment.get(PropertiesUtil.java:513)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.util.PropertiesUtil$Environment.access$200(PropertiesUtil.java:434)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.util.PropertiesUtil.getStringProperty(PropertiesUtil.java:382)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.util.PropertiesUtil.getBooleanProperty(PropertiesUtil.java:169)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.status.StatusLogger.isDebugPropertyEnabled(StatusLogger.java:143)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.status.StatusLogger.isEnabled(StatusLogger.java:426)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.status.StatusLogger.isEnabled(StatusLogger.java:354)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1914)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.spi.AbstractLogger.debug(AbstractLogger.java:463)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.PatternProcessor.formatFileName(PatternProcessor.java:291)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.PatternProcessor.formatFileName(PatternProcessor.java:278)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.AbstractRolloverStrategy.getEligibleFiles(AbstractRolloverStrategy.java:94)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.AbstractRolloverStrategy.getEligibleFiles(AbstractRolloverStrategy.java:87)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.DefaultRolloverStrategy.rollover(DefaultRolloverStrategy.java:524)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.RollingFileManager.rollover(RollingFileManager.java:504)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.RollingFileManager.rollover(RollingFileManager.java:394)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.rolling.RollingFileManager.checkRollover(RollingFileManager.java:308)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.appender.RollingFileAppender.append(RollingFileAppender.java:300)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:161)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:134)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:125)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:89)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:675)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:633)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:616)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:552)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:82)
    at org.apache.logging.log4j.core@8.5.2/org.apache.logging.log4j.core.Logger.log(Logger.java:161)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2106)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.internal.DefaultLogBuilder.logMessage(DefaultLogBuilder.java:234)
    at org.apache.logging.log4j@2.18.0/org.apache.logging.log4j.internal.DefaultLogBuilder.log(DefaultLogBuilder.java:162)
    at org.apache.poi.hdgf.chunks.Chunk.processCommands(Chunk.java:184)
    at org.apache.poi.hdgf.chunks.ChunkFactory.createChunk(ChunkFactory.java:207)
    at org.apache.poi.hdgf.streams.ChunkStream.findChunks(ChunkStream.java:66)
    at org.apache.poi.hdgf.streams.PointerContainingStream.findChildren(PointerContainingStream.java:70)
    at org.apache.poi.hdgf.streams.PointerContainingStream.findChildren(PointerContainingStream.java:77)
    at org.apache.poi.hdgf.streams.PointerContainingStream.findChildren(PointerContainingStream.java:77)
    at org.apache.poi.hdgf.HDGFDiagram.<init>(HDGFDiagram.java:89)
    at org.apache.poi.hdgf.extractor.VisioTextExtractor.<init>(VisioTextExtractor.java:52)
    at org.apache.tika.parser.microsoft.OfficeParser.parse(OfficeParser.java:232)
    at org.apache.tika.parser.microsoft.OfficeParser.parse(OfficeParser.java:175)
    at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298)
    at org.apache.tika.parser.AutoDetectParser.parse(AutoDetectParser.java:180)
    at org.apache.tika.parser.DelegatingParser.parse(DelegatingParser.java:71)
    at org.apache.tika.extractor.ParsingEmbeddedDocumentExtractor.parseEmbedded(ParsingEmbeddedDocumentExtractor.java:108)
    at org.apache.tika.parser.microsoft.ooxml.AbstractOOXMLExtractor.handleEmbeddedFile(AbstractOOXMLExtractor.java:406)
    at org.apache.tika.parser.microsoft.ooxml.AbstractOOXMLExtractor.handleEmbeddedOLE(AbstractOOXMLExtractor.java:351)
    at org.apache.tika.parser.microsoft.ooxml.AbstractOOXMLExtractor.handleEmbeddedPart(AbstractOOXMLExtractor.java:267)
    at org.apache.tika.parser.microsoft.ooxml.AbstractOOXMLExtractor.handleEmbeddedParts(AbstractOOXMLExtractor.java:217)
    at org.apache.tika.parser.microsoft.ooxml.AbstractOOXMLExtractor.getXHTML(AbstractOOXMLExtractor.java:138)
    at org.apache.tika.parser.microsoft.ooxml.OOXMLExtractorFactory.parse(OOXMLExtractorFactory.java:242)
    at org.apache.tika.parser.microsoft.ooxml.OOXMLParser.parse(OOXMLParser.java:115)
    at org.apache.tika.parser.ParserDecorator.parse(ParserDecorator.java:152)
    at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298)
    at org.apache.tika.parser.AutoDetectParser.parse(AutoDetectParser.java:180)
    at org.apache.tika.Tika.parseToString(Tika.java:525)
    at org.elasticsearch.ingest.attachment.TikaImpl.lambda$parse$0(TikaImpl.java:97)
```

Reply:
That's coming from https://github.com/elastic/elasticsearch/issues/91964 (and not what I saw before for this issue, which was slf4j related). We fixed that previously by forcing static init of the Unbox class, but it looks like it didn't stick.

I'm thinking now we do need another fix at a lower level in our log4j setup. I would like to avoid giving env access to all Tika processors. I created https://github.com/elastic/elasticsearch/issues/93714 to track this.